spec/unit/logstash_output_spec.rb

```diff
@@ -184,10 +190,100 @@
+
+        expect(registered_plugin).to receive(:response_success?).exactly(2).and_call_original
+        expect(registered_plugin).to receive(:log_response).once
+        expect(registered_plugin).to receive(:retryable_exception?).never
+        expect(registered_plugin).to receive(:log_exception).never
+
+        registered_plugin.multi_receive([event])
+      end
+
+      it "doesn't retry on other non-retriable errors" do
+        allow(response).to receive(:code).and_return([400, 404].sample)
+        allow(response).to receive(:body).and_return("Non-retriable error.", "Send succeeded.")
+
+        expect(registered_plugin).to receive(:response_success?).once.and_call_original
+        expect(registered_plugin).to receive(:log_response).once
+        expect(registered_plugin).to receive(:retryable_exception?).never
+        expect(registered_plugin).to receive(:log_exception).never
+
+        registered_plugin.multi_receive([event])
+
+      end
+    end
+
+    describe "with an exception message" do
+      let(:config) { super().merge("hosts" => %w[127.0.0.1 my-ls-downstream.com:1234]) }
+      let(:exception_raising_client) { double("Exceptional Manticore client") }
+
+      before do
+        # a simulation, where 127.0.0.1 host raises an exception and retry to my-ls-downstream.com will be succeeded
+        allow(registered_plugin.http_client).to receive(:post) do | url, _, _|
+          url.eql?("https://127.0.0.1:9800") ? exception_raising_client : client
+        end
+      end
+
+      it "with Manticore socket timeout" do
+        allow(exception_raising_client).to receive(:call).and_raise(Manticore::SocketTimeout.new)
+        allow(client).to receive(:call).and_return(response)
+        allow(response).to receive(:code).and_return(200)
+
+        # retry succeeds and breaks loop
+        expect(registered_plugin).to receive(:retryable_exception?).once.and_call_original
+        expect(registered_plugin).to receive(:log_exception).once
+        expect(registered_plugin).to receive(:response_success?).once.and_call_original
+        expect(registered_plugin).to receive(:log_response).never
+
+        registered_plugin.multi_receive([event])
+      end
+
+      it "with connection reset by peer error" do
+        allow(exception_raising_client).to receive(:call).and_raise(Manticore::UnknownException.new "Connection reset by peer")
+        allow(client).to receive(:call).and_return(response)
+        allow(response).to receive(:code).and_return(200)
+
+        expect(registered_plugin).to receive(:response_success?).once.and_call_original # second loop succeeds and break
+        expect(registered_plugin).to receive(:log_response).never
+        expect(registered_plugin).to receive(:retryable_exception?).once.and_call_original
+        expect(registered_plugin).to receive(:log_exception).once
+
+        registered_plugin.multi_receive([event])
+      end
+    end
+  end
```

Comment:
We're mixing rspec paradigms here, and as-written we are very tightly-coupled with the implementation.

Let's sync on a call today to find a path forward.